src/protocol/sort/generate_permutation.rs

```diff
@@ -174,19 +203,143 @@ pub async fn generate_permutation_and_reveal_shuffled<F: Field>(
     num_bits: u32,
 ) -> Result<RevealedAndRandomPermutations, Error> {
     let sort_permutation = generate_permutation(ctx.narrow(&SortKeys), sort_keys, num_bits).await?;
-    shuffle_and_reveal_permutation(
+    Ok(shuffle_and_reveal_permutation(
         ctx.narrow(&ShuffleRevealPermutation),
         u32::try_from(sort_keys[0].len()).unwrap(),
         sort_permutation,
+        None,
     )
     .await
+    .unwrap()
+    .reveal_and_random_permutation)
+}
+
+#[allow(dead_code)]
+#[embed_doc_image("malicious_sort", "images/sort/malicious-sort.png")]
+/// Returns a sort permutation in a malicious context.
+/// This runs sort in a malicious context. The caller is responsible to validate the accumulater contents and downgrade context to Semi-honest before calling this function
+/// The function takes care of upgrading and validating while the sort protocol runs.
+/// It then returns a semi honest context with output in Replicated format. The caller should then upgrade the output and context before moving forward
+///
+/// Steps
+/// 1. [Malicious Special] Upgrade the context from semihonest to malicious and get a validator
+/// 2. [Malicious Special] Upgrade 0th sort bit keys
+/// 3. Compute bit permutation that sorts 0th bit
+///
+/// For 1st to N-1th bit of input share
+/// 1. i. Shuffle the i-1th composition
+///   ii. [Malicious Special] Validate the accumulator contents
+///  iii. [Malicious Special] Malicious reveal
+///   iv. [Malicious Special] Downgrade context to semihonest
+/// 2. i. [Malicious Special] Upgrade ith sort bit keys
+///   ii. Sort ith bit based on i-1th bits by applying i-1th composition on ith bit
+/// 3. Compute bit permutation that sorts ith bit
+/// 4. Compute ith composition by composing i-1th composition on ith permutation
+/// In the end, following is returned
+///    i. n-1th composition: This is the permutation which sorts the inputs
+///   ii. Validator which can be used to validate the leftover items in the accumulator
+///
+/// ![Malicious sort permutation steps][malicious_sort]
+/// # Panics
+/// If sort keys dont have num of bits same as `num_bits`
+/// # Errors
+pub async fn malicious_generate_permutation<'a, F>(
+    ctx: SemiHonestContext<'a, F>,
+    sort_keys: &[Vec<Replicated<F>>],
+    num_bits: u32,
+) -> Result<(MaliciousValidator<'a, F>, Vec<MaliciousReplicated<F>>), Error>
+where
+    F: Field,
+{
+    let mut malicious_validator = MaliciousValidator::new(ctx);
+    let mut m_ctx = malicious_validator.context();
+    let m_ctx_0 = m_ctx.narrow(&Sort(0));
+    assert_eq!(sort_keys.len(), num_bits as usize);
+
+    let upgraded_sort_keys = m_ctx.upgrade_vec(0, sort_keys[0].clone()).await?;
```

Comment:
I think that maybe you want to use a step name for the upgrades so that you don't have to continually update the starting number.  That means using a version based on `upgrade_with()`.